File: cpp/leetcode/202.cpp
class Solution {
  public:
    bool isHappy(int n) {
        int fast = n, slow = n;
        do {
            slow = getSum(slow);
            fast = getSum(fast);
            fast = getSum(fast);
        } while (fast != slow);
        return slow == 1;
    }

  private:
    int getSum(int n) {
        int res = 0;
        while (n) {
            res += pow(n % 10, 2);
            n /= 10;
        }
        return res;
    }
};

File: python/leetcode/202.py
class Solution:
    def isHappy(self, n):
        res = set()
        while True:
            res.add(n)
            n = sum([int(x) * int(x) for x in str(n)])
            if n == 1 or n in res:
                break
        return n == 1


class Solution:
    def isHappy(self, n):
        slow, fast = n, n
        while True:
            slow = sum([int(x) * int(x) for x in str(slow)])
            fast = sum([int(x) * int(x) for x in str(fast)])
            fast = sum([int(x) * int(x) for x in str(fast)])
            if fast == slow:
                break
        return slow == 1
